lib/Conversion/HWArithToHW/HWArithToHW.cpp

```diff
@@ -299,88 +349,69 @@ struct BinaryOpLowering : public OpConversionPattern<BinOp> {
   }
 };
 
-template <class TOp>
-struct ArgResOpConversion : public OpConversionPattern<TOp> {
-  // Generic pattern which replaces an op by one of the same type, but with
-  // converted operands and result types.
-  using OpConversionPattern<TOp>::OpConversionPattern;
-  using OpAdaptor = typename OpConversionPattern<TOp>::OpAdaptor;
-
-  // Function that will replace the op with a type-converted new op.
-  using ReplacementFunc = std::function<void(
-      ConversionPatternRewriter &, TOp, llvm::SmallVector<Type>, OpAdaptor &)>;
+struct TypeConversionPattern : public ConversionPattern {
+public:
+  TypeConversionPattern(MLIRContext *context, TypeConverter &converter)
+      : ConversionPattern(converter, MatchAnyOpTypeTag(), 1, context) {}
+  using ConversionPattern::ConversionPattern;
 
-  ArgResOpConversion(TypeConverter &tc, MLIRContext *ctx, ReplacementFunc f)
-      : OpConversionPattern<TOp>(tc, ctx), replacementFunc(f) {}
+  // Generic pattern which replaces an operation by one of the same type, but
+  // with converted operands and result types. Uses generic builders based on
+  // OperationState to make sure that this pattern can apply to _any_ operation
+  // which uses HWArith types.
 
   LogicalResult
-  matchAndRewrite(TOp op, OpAdaptor adaptor,
+  matchAndRewrite(Operation *op, ArrayRef<Value> operands,
                   ConversionPatternRewriter &rewriter) const override {
-    llvm::SmallVector<Type> convResTypes;
-    if (failed(this->getTypeConverter()->convertTypes(op->getResultTypes(),
-                                                      convResTypes)))
-      return failure();
-    replacementFunc(rewriter, op, convResTypes, adaptor);
+    llvm::SmallVector<NamedAttribute, 4> newAttrs;
+    newAttrs.reserve(op->getAttrs().size());
+    for (auto attr : op->getAttrs()) {
```

Comment:
Do we really want to convert all type attributes? Maybe just the inherent (non-dialect) attributes would be safer?